tutorials/continuous-deployment-pipeline-google-container-engine-with-codeship/index.md

```diff
@@ -0,0 +1,270 @@
+---
+title: Continuous Deployment Pipeline to Google Container Engine using Codeship
+description: Learn how to create a continuous deployment pipeline to Container Engine from Codeship.
+author: kellyjandrews
+tags: CD, Container Engine, Codeship, Pipeline
+date_published: 2017-08-28
+---
+This tutorial explains how to create a Continuous Deployment
+pipeline to Google Container Engine using Codeship. You will
+learn how to deploy a containerized application when new code is merged into the
+master branch and all integration tests have passed.
+
+Continuous Deployment is the practice of automating the delivery of software to
+production after passing a series of tests. Although not for everyone, it should
+be the goal of any software development and DevOps team not restricted by
+regulatory constraints.
+
+## Before you begin
+
+Take the following steps to enable the Google Container Engine API:
+
+1.  Visit the [Container Engine](https://console.cloud.google.com/projectselector/kubernetes)
+    page in the Google Cloud Platform Console.
+1.  Create or select a project.
+1.  Wait for the API and related services to be enabled, which can take several
+    minutes.
+1.  [Enable billing][billing] for your project.
+
+[billing]: https://support.google.com/cloud/answer/6293499#enable-billing
+
+Make sure you have the following:
+
+1.  [Docker CE Version 17.03.01](https://www.docker.com/community-edition)
+1.  [Codeship Account](https://app.codeship.com/registrations/new)
+1.  [Codeship Jet CLI](https://documentation.codeship.com/pro/builds-and-configuration/cli/)
+1.  [`hello-express` source code](https://github.com/kellyjandrews/hello-express)
+1.  [GitHub](https://github.com/), [Bitbucket](https://bitbucket.org/), or [GitLab](https://about.gitlab.com/) account
+
+## Setting up the continuous deployment pipeline
+
+### Step 1: Create service account
+
+The interactions with the Google Cloud Platform API from Codeship require a service
+account with permissions to the Cloud Storage and Container Engine services.
+
+Follow these steps to create a service account:
+
+1.  Visit the [Service accounts](https://console.cloud.google.com/projectselector/iam-admin/serviceaccounts)
+    page in the Google Cloud Platform Console.
+1.  Select your project.
+1.  Click `Create Service Account`.
+1.  Enter a name.
+1.  Click `Select a Role` and choose the following permissions:
+
+    * Project &rarr; Service Account Actor
+    * Container &rarr; Container Engine Admin
```

Comment:
have you been able to figure out if you really need Container Engine "Admin"?

Reply:
Yes - the cluster creation and deployment creation require the admin level.

Reply:
@kellyjandrews I'm confused, does CodeShip create the cluster? Deployments via kubectl only require `Container Engine Developer` role.

Reply:
The initial build creates the cluster and the deployment. Essentially everything completed in the "deploy containerized app" tutorial runs in the first script. 

The removal script, I would assume, also requires admin level.

Is it best practice to not create an admin level service worker?

Reply:
@ahmetb any additional feedback? How should I proceed from here?

Reply:
It's not a good practice to give a CI/CD system `container.admin` permission as it grants full access to all clusters. `container.developer` still can use `kubectl` to deploy/delete Kubernetes resources to the cluster.

I think if you tell people to create the cluster as a `Step 0` it would be way better. It's a very basic command anyway:

    gcloud container clusters create hello --num-nodes=3

This way users can learn about how to create a cluster and they will maybe run `--help` to learn about more configurable knobs. By hiding this command into your script, your gains will be minimal.

It's definitely not a good practice to create a cluster in your CI/CD system (unless you're creating/tearing down a cluster for every build, which I'm assuming not the case here).

Reply:
(Sorry that comment was on my browser for the last 20 hours, I apparently forgot to hit send)

Reply:
Ok. Makes sense. I'll work that up. Everything else would remain the same then for the initial set up. I can also remove the clean up script as well and use the standard instructions there as well.

I will  also adde requirement for GCloud sdk installed as well.